src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala

```diff
@@ -824,6 +824,8 @@ abstract class ClassfileParser(reader: ReusableInstance[ReusableDataReader]) {
           val arg = Literal(Constant("see corresponding Javadoc for more information."))
           sym.addAnnotation(DeprecatedAttr, arg, Literal(Constant("")))
           in.skip(attrLen)
+          if (sym == clazz)
+            staticModule.addAnnotation(DeprecatedAttr, arg, Literal(Constant("")))
```

Comment:
It strikes me as odd that we need to annotate the `ModuleSymbol` (`staticModule`) rather than the `ModuleClassSymbol` (`staticModule.moduleClass` or equivalently `ClassfileParser.this.moduleClass`). The `ModuleClassSymbol` is the owner of the static methods, and the check for enclosing `@deprecated` looks up the owner chain:

```scala
if (sym.isDeprecated && !currentOwner.ownerChain.exists(x => x.isDeprecated || x.hasBridgeAnnotation))
        currentRun.reporting.deprecationWarning(pos, sym)
```

But your tests cases passing suggest that I'm wrong...

Reply:
Oh, we're reporting the `p1.DeprecatedClass` as deprecated, not the call to `foo` which that qualifies.